File: binary search/Search in Rotated Sorted Array.cs
public class SolutionSearch
{
    public int Search(int[] nums, int target)
    {
        int l = 0, h = nums.Length - 1;

        while (l <= h)
        {
            int m = l + (h - l) / 2;

            if (nums[m] == target)
                return m;

            if (nums[m] <= nums[h]) // right side is in order
            {
                if (target > nums[m] && target <= nums[h])
                    l = m + 1;
                else
                    h = m - 1;
            }
            else // left side is in order
            {
                if (target < nums[m] && target >= nums[l])
                    h = m - 1;
                else
                    l = m + 1;
            }
        }
        return -1;
    }
}

File: binary search/Search in Rotated Sorted Array.py
def search(nums: list[int], target: int) -> int:
    l, r = 0, len(nums) - 1
    
    while l <= r:
        mid = l + (r - l) // 2

        if nums[mid] == target:
            return mid

        if nums[l] <= nums[mid]:
            if nums[l] <= target < nums[mid]:
                r = mid - 1
            else:
                l = mid + 1
        else:
            if nums[mid] < target <= nums[r]:
                l = mid + 1
            else:
                r = mid - 1

    return -1

print(search([4,5,6,7,0,1,2], 0))
print(search([4,5,6,7,0,1,2], 3))
